### ISSUE TITLE
Produce ExposedThing from a ThingModel instance

### ISSUE THREAD

**Issue by relu91** [w3c/wot-scripting-api#304]
Now that a new paragraph about Thing Model is landed in the TD specification (https://github.com/w3c/wot-thing-description/pull/1024) we should evaluate possible choices about how to integrate it in the Scripting API. 

As I see it, a good starting point for integrating Thing Models is in the workflow for producing `ExposedThing`s. We could exploit the work done with `ExposedThingInit` and define a transformation algorithm between a TM instance and the `ExposedThingInit` object. We could even change the `produce` method signature such as it can handle `ThingModel` or  `ExposedThingInit` instances as arguments. 

Another option is to use the algorithm described in the mentioned PR. But as I raised during the call it is a bit vague when looking at it from an implementer's point of view. The TD group would like to keep it as it is since they would like to have a process that could be used also by tools or even humans rather than an automatic process computed by a runtime. 

Opinions?

**Comment by zolkis** [comment]
We'd define ThingModel as object as well...

I think we should still separate the notion of "init object" as a JS concept from what it can map (externally defined notions/objects) in the algorithms.
After that PR, the init object for produce() can be a ThingModel or a ThingDescription (coming from another ConsumedThing), or any JS object that passes validation (as defined in the algorithm), regardless how was it put together.

**Comment by relu91** [comment]
> After that PR, the init object for produce() can be a ThingModel or a ThingDescription (coming from another ConsumedThing), or any JS object that passes validation (as defined in the algorithm), regardless how was it put together.

The point is that we need some processing steps for ThingModel because it can have placeholders like the following:
```js
{
   title:"{{YOURTITLE}}"
}
```
Without mentioning the fact that it might need to be merged with other bases TMs. So we can't use it as it is. On the other hand, you're right about TDs since they are also valid PartialTD (or `ExposedThingInit`).

**Comment by zolkis** [comment]
Just to be exact, I make a difference between `ExposedThingInit` (a JS object that is undergoing the validation algorithm defined in the Scripting API spec) and any terms defined in the TD TF, including Partial TD, TD, Thing Model etc. 

The normative steps are in the produce() and expose() steps, and in the sub-algorithms they use, like validation.

It is out of scope for Scripting how does someone merge Thing Models, how the TD or Partial TD was generated etc.

It is in scope for Scripting to handle the use cases that are legal for non-Scripting use cases, i.e. network-interface for WoT (which used to be called WoT Interface or WoT Network Interface, though today it might have different name).

Since exposing Things is not standardized at WoT Interface level, the Scripting API defines its behavior. Therefore there is no external requirement on what and how to accept. Which is why it makes sense to me to syntactically separate the init object from terms defined outside of Scripting, and semantically map them in the algorithms.

**Comment by relu91** [comment]
> Just to be exact, I make a difference between ExposedThingInit (a JS object that is undergoing the validation algorithm defined in the Scripting API spec) and any terms defined in the TD TF, including Partial TD, TD, Thing Model etc.

> The normative steps are in the produce() and expose() steps, and in the sub-algorithms they use, like validation.

Yes, I think we are on the same page. Still, as you mentioned here: 
> We'd define ThingModel as object as well...

We have to add a new JS object that represents a ThingModel instance and refer to an algorithm to transform it to an `ExposedThingInit`. As said also in https://github.com/w3c/wot-thing-description/issues/1047 it would be ideal to have this algorithm defined in TD doc and we just use it. I think there're strong use cases to have TMs as first citizens in a Script runtime.

**Comment by zolkis** [comment]
The way I think about it is that `produce()` accepts an _object_ (which we call `ExposedThingInit` in Web IDL) that is undergoing validation. This is the function signature, that we accept an object. All the rest are external notions to Scripting and are just creating confusion IMHO. But at _validation level_ we can decide what _source_ to check: ThingModel? Thing Description? Partial TD? ConsumedThing?  

In other words, I would not bother to represent all the possible sources or mappings (Thing Model, Thing Description, Partial TD, ConsumedThing) on Web IDL signature level. It just complicates things, since Web IDL is an intermediate representation that we don't even use for generating any code, we just use Web IDL as _prose_ in the spec, to illustrate the API in a standard way. 

On validation (algorithm) level, we can just make clauses to handle specific TD stanza (title, etc), which may have various sources (Thing Model, etc).

**Comment by relu91** [comment]
This is a duck typing approach. Accept everything and then validate which is the standard way to build JS based applications. However, I feel like with WebIDL we could convey more hints to the implementers. We should still validate the input object but at least everything it's clear from the method signature. Which downside do you foresee if we'll define a ThingModel object?

Maybe it's me because I really like type systems 🤣

**Comment by danielpeintner** [comment]
Q: I wonder whether we actually need (should?) go along that path. According to my understanding I would rather start from TD only. There might be implementations offering ThingModel to TD conversion but do we really need to cover that in Scripting API?

Comments welcome!

FYI: implementation-wise there seems to be already support, see https://www.npmjs.com/package/json-placeholder-replacer

**Comment by zolkis** [comment]
> While the idea of having a separation of concern is good, I would expect that the architectural concepts are reported as runtime objects in the API. 

If we can define a Web IDL _interface_ for ThingModel, ThingDescription (, ConsumedThing), then we can _OR_ the types.

If all of these are defined as generic _objects_, then I would use the idiomatic xxxTypeInit dictionaries that all Web specs are using and map to the source types in the algorithms. That is also more future-proof :).

**Comment by relu91** [comment]
> Q: I wonder whether we actually need (should?) go along that path. According to my understanding I would rather start from TD only. There might be implementations offering ThingModel to TD conversion but do we really need to cover that in Scripting API?

I think this is the same answer of "Why do we need an ExposeThingInit?". The thing is that a TD describes a fully running WebThing instance and it cannot be used* to instruct a runtime to _produce_ a WebThing. It is way more practical to use ExposeThingInit. What would be more useful is that ExposeThingInit objects could be shared and extended in other projects to create a nice structure of dependencies. This feature is exactly what a ThingModel provides. Instead of defining our own rules for extending ExposeThingInit instances, we can leverage the concept of ThingModel and enjoy its features (like placeholders, and a standard serialization/sharing format). That's why I opened this issue. I hope it answers your question!

* side note: you could use a TD but all the forms and security definitions should comply with the current runtime implementation which in the end means that you should know  at dev time the capabilities of your deployment system (not really common)

> If we can define a Web IDL interface for ThingModel, ThingDescription (, ConsumedThing), then we can OR the types.

Yep, that's what I'd do.

**Comment by zolkis** [comment]
> Yep, that's what I'd do. 

I think that should be a spec of its own. Not sure it should be hosted by the TD TF, or by the Scripting TF, but it should be a proper web spec, with Web IDL, algorithms, all the regular sections etc. I see that an ideal solution, but until we have that, IMHO we should stick with the ExposedThingInit dictionary and express the ThingModel knowledge in the algorithms (instead in syntax which is missing if we don't have a rigorously defined interface for a ThingModel). 

This topic should be discussed in the F2F.

**Comment by danielpeintner** [comment]
Note: I think we need to be aware that ThingModel adds a lot more complexity. A TD (or ExposedThingInit) is a _single_ self-contained file.

A TD based on ThingModels can be build upon _several_ TMs
* one for example covering the "light" part, the other the "switch" part
* one extending constructs of the other (causing linking/reference TM issues)
* ... combinations ...

see https://w3c.github.io/wot-thing-description/#thing-model

**Comment by zolkis** [comment]
> A TD based on ThingModels can be build upon several TMs.

True. IMHO how to assemble TM fragments into an ExposedThingInit should be out of scope for the Scripting API (`produce()`).
It should be handled by a separate tool (or separate API), since it will have to solve separate problems.
IMHO the produce() function should accept only one init object (a single TM, TD or ConsumedThing).
How to arrive to that init object is the responsibility of another tool, or API.
One more reason why we should keep the separation of concerns and go with ExposedThingInit for `produce()`.

**Comment by relu91** [comment]
Notice that ThingModels have placeholders. I foresee that a really common use case for those placeholders is to fill them *at runtime* not statically by some tool. We don't have placeholders in `ExposedThingInit`. If we are leaving out ThingModels from the API, I think that we should at least provide support for placeholders.

**Comment by zolkis** [comment]
ExposedThingInit is an _object_. It may contain whatever the runtime can interpret, but we'd better specify its content.

The use case in Scripting is that the produce() API accepts one document (Thing Description) based upon which a runtime can construct an exposed Thing with its interactions (hooks to be provided via the ExposedThing API).

How we arrive to the TD provided to produce() is out of scope for produce(). 
I hope we don't need to specify a JSON editor.

**Comment by relu91** [comment]
> ExposedThingInit is an object. It may contain whatever the runtime can interpret, but we'd better specify its content.

So can't we specify how to use placeholder if they are as content in the ExposedThingInit?

> I hope we don't need to specify a JSON editor.

Of course not. 🤣 

> The use case in Scripting is that the produce() API accepts one document (Thing Description) based upon which a runtime can construct an exposed Thing with its interactions (hooks to be provided via the ExposedThing API).

Then I think we should improve this use case. If the produce method should accept only TDs then it should validate them with the rules defined in the TD spec. This means that the object passed as the argument of the produce method should contain every form and every href. We all know that is pretty limiting because the application should know the supported features of the runtime. The real use case IMHO is the following:

The produce API accepts a hint for the generation of a valid Thing Description. The hint is used as a skeleton for the produced Thing Description which will be based on specific runtime characteristics (i.e., supported protocols, security schemas etc.). The resulting ExposedThing will have the generated TD and interaction hooks. 

The hint is the newly defined ExposedThingInit.

**Comment by zolkis** [comment]
All input to `produce()` is _hint_ to the implementation, exactly to match the local runtime capabilities. This is already reflected in our algorithms.

So if a TD is provided (which originally was meant for consumption), the impl will distill the interaction names, types etc and will fill the forms based on local runtime. A TD provided to produce() is just used for input, it is the `expose()` method that generates the new TD, which will have the same interactions but different forms from the input.

Similarly, if the input is a ThingModel or a ConsumedThing, the current algorithms just take from them the interaction names, arguments, types etc. Later the app adds the handler functions, then calls expose() which will actually launch the main work, once it has the interactions and the handlers. 

Certainly, the details how to actually generate bindings are not defined in our algorithms yet. That is private to the implementations at the moment.

Anyway, the point is that ThingModel is a new and incomplete development. We are lacking a spec for ThingModel, algorithms for placeholders, a Web IDL interface for ThingModel, etc. 

Therefore I still stand by the design rationale to separate the concerns and map/handle the input of produce() in algorithms, rather than in Web IDL syntax. We don't have the prerequisites to be able to syntactically describe all these inputs. 

In the first phase, we should keep the current API and improve the produce() and expose() algorithms wrt how to handle input (including Thing Models).

That can handle multiple Thing Model input as well: prior to calling produce(), the app needs to fetch and parse various Thing Model sources, and assemble them into the same JS object that would be input to produce().

IMHO it is not for produce() to fetch and parse those Thing Models.

**Comment by relu91** [comment]
Ok so I think we are reaching the following consensus: 
We want to keep produce lean and focused on the single task of produce an ExposedThing from a hint object. This object can be anything as long as it follows the validation algorithm that we drafted in the current spec. Consequently, the produce method is flexible enough to possibly handle the output of other libraries/tools that can transform a ThingModel into the hint object. 

This solution is optimal because it does not add any new constraint to the API. While I see this as a good solution, thinking about an actual developer you receive a couple of TM as files and wants just to create a WebThing from them quite counter-intuitive. 

>There can be an alternative call for creating TDs out of TMs, placeholder maps et cetera. Having said that, I am not sure if this should be part of the core API.

Would defining another API document just for TM runtime handling be a little bit overkill? Or should we completely ignore TMs and leave to 3rd part tools and libraries?

**Comment by zolkis** [comment]
> Would defining another API document just for TM runtime handling be a little bit overkill? Or should we completely ignore TMs and leave to 3rd part tools and libraries?

The latter. Though we do care about TMs, but it should be well specified first. Also, TM manipulation is out of context for produce(), which should only be concerned about having a correct algorithm on how to deal with input (the hint object). IoW, as said above, all the knowledge is in the algorithm at the moment, and we don't map it out to syntactic level until all of it is well specified.

**Comment by danielpeintner** [comment]
I agree also, we should _re-consider_ our next steps once the ThingModel has been settled.

A possible solution I see is to have yet another [conformance class](https://w3c.github.io/wot-scripting-api/#the-wot-namespace).
At the moment we have:
* WoT Consumer 
*  WoT Producer
*  WoT Discovery

The **new**  conformance class could deal with handling TMs which then could be used elsehwere.

**Comment by zolkis** [comment]
> The new conformance class could deal with handling TMs which then could be used elsewhere.

I agree. But we could also explore other possibilities, like extending an ExposedThing object with another ExposedThing (i.e. merging into the current one, where we define how merging is done). This could be part of the ExposedThing API.
This would be quite trivial, the question is what to do on conflict (as with merges in general).

```JS
partial interface ExposedThing {
    Promise<undefined> merge(ExposedThing input, optional MergeOptions options = "strict");
};

enum MergeOptions { 
    "strict", // error on every conflict
    "override", // input replaces conflicting items
    "try", // on conflict, keep the old items
};
```

**Comment by relu91** [comment]
I agree let's postpone the discussion! Merging seems nice but I would go with another conformance class or dedicated functions. Anyhow, let's discuss it when the TM definition is more settled.

**Comment by zolkis** [comment]
> I would go with another conformance class or dedicated functions 

Certainly. We can draft a variety of API proposals, then discuss and choose.
For the record, even an extension (partial interface) can be in a separate conformance class.
